src/components/inputs/BasicTextField/BasicTextField.tsx

```diff
@@ -1,32 +1,75 @@
 import TextField from '@mui/material/TextField';
-import {Control, Controller} from 'react-hook-form';
+import {HTMLInputTypeAttribute, useState} from 'react';
+import {Control, Controller, FieldError, FieldValues, Path, RegisterOptions} from 'react-hook-form';
+import PasswordAdornment from '../PasswordAdornment/PasswordAdornment';
 
-interface BasicTextFieldProps {
-  name: string;
-  control: Control<any>;
-  label: string;
-  rules?: Record<string, any>;
+interface FormInputControllerProps<FieldsType extends FieldValues> {
+  name: Path<FieldsType>;
+  defaultValue?: string;
+  rules?: RegisterOptions;
+  error?: FieldError;
+  control: Control<FieldsType>;
 }
 
-const BasicTextField: React.FC<BasicTextFieldProps> = ({name, control, label, rules}) => {
+interface Props<FieldsType extends FieldValues> extends FormInputControllerProps<FieldsType> {
+  label?: string;
+  className?: string;
+  type?: HTMLInputTypeAttribute;
+}
+
+const BasicTextField = <FieldsType extends FieldValues>({name, control, type = 'text', label, rules, className}: Props<FieldsType>) => {
+  const [showPassword, setShowPassword] = useState<boolean>(false);
+
+  const handleClickShowPassword = (): void => {
+    setShowPassword((show) => !show);
+  };
+
+  const handleMouseDownPassword = (event: React.MouseEvent<HTMLButtonElement>): void => {
+    event.preventDefault();
+  };
+
+  const setType = (): string => {
+    if (type === 'password') {
+      return showPassword ? 'text' : 'password';
+    } else {
+      return type;
+    }
+  };
+
   return (
-    <Controller
-      name={name}
-      control={control}
-      rules={rules}
-      render={({field: {onChange, value}, fieldState: {error}, formState}) => (
-        <TextField
-          helperText={error ? error.message : null}
-          size="small"
-          error={!!error}
-          onChange={onChange}
-          value={value || ''}
-          fullWidth
-          label={label}
-          variant="outlined"
-        />
-      )}
-    />
+    <div className={className}>
+      <Controller
+        name={name}
+        control={control}
```

Comment:
Please tell me what is the type of control?

Reply:
```
const control: Control<{
    phoneNumber?: string | undefined;
    email: string;
    password: string;
    firstName: string;
    lastName: string;
    confirmPassword: string;
}, any>
```